core/unix/module_elf.c

```diff
@@ -253,14 +301,28 @@ module_fill_os_data(ELF_PROGRAM_HEADER_TYPE *prog_hdr, /* PT_DYNAMIC entry */
             int soname_index = -1;
             char *dynstr = NULL;
             size_t sz = mod_max_end - mod_base;
+            size_t mapped_size = 0;
             while (dyn->d_tag != DT_NULL) {
                 if (dyn->d_tag == DT_SONAME) {
                     soname_index = dyn->d_un.d_val;
                     if (dynstr != NULL)
                         break;
                 } else if (dyn->d_tag == DT_STRTAB) {
-                    dynstr = (char *)elf_dt_abs_addr(dyn, base, sz, view_size, load_delta,
-                                                     at_map, dyn_reloc);
+                    /* i#3385 Some ELF files' .dynstr section occurs late in the
+                     * file. This requires use of the file offset rather than the
+                     * virtual offset because the .dynstr segment has not yet been
+                     * mapped. at_map indicates whether all segments are fully
+                     * loaded (at_map=false => use virtual offsets) or file just
+                     * at the first flat mmap (at_map=true => use file offsets).
+                     */
+                    if (at_map == 1) {
+                        mapped_size = module_map_file(out_data);
```

Comment:
Should we check the app's map size and if it's big enough (I guess we can only check vs the file size since we don't have the dynstr offset here), not bother to do our own map?  OK to not do that for code simplicity but I think it's worth a comment that in some sense we're being wasteful with an extra map in those cases.

Reply:
I've opted for a XXX comment, for now.